Homework for sprint №14

sobolevkir opened the pull request:
Всё поменял. Теперь вопрос ещё по блоку кода:

`val track = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            intent.getParcelableExtra(CURRENT_TRACK, Track::class.java)
        } else intent.getParcelableExtra(CURRENT_TRACK)`
        
Студия обращает внимание: `'getParcelableExtra(String!): T?' is deprecated. Deprecated in Java`
Что в этом случае лучше сделать?
Я хотел сделать проверку, т.к. конструктор метода getParcelableExtra в разных API разный)


И ещё: я постарался учесть замечания по поводу переключения темы. Ранее наставник писал, что в рамках данного приложения переключение темы зависит только от положения свича и не зависит от темы устройства. Но Я всё-равно сделал проверку, что если нет записи в SharedPrefs, то тему включаем соответственно теме устройства. А уже если пользователь переключил тему, то с этого момента она будет соответствовать настройке. Правильно ли я понял логику?

sobolevkir committed:
Использован ViewBinding для SettingsActivity

sobolevkir committed:
Изменения:
1) В поиск добавлен Debounce
2) Добавлено проигрывание превью треков

IliaAnK commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/TrackListAdapter.kt:
```diff
@@ -1,30 +1,47 @@
+            binding.root.setOnClickListener {
+                if (clickDebounce()) {
+                    onTrackClickListener?.onTrackClick(tracks[position])
+                    val playerIntent = Intent(it.context, PlayerActivity::class.java)
```
Рекомендация. Шаг 1:
Не стоит запускать `Activity` внутри `onBindViewHolder`. 
Здесь нужно только уведомить `Activity/Fragment` о том, что на элемент нажали.

Есть несколько способов сделать это:
1. Это листнер. В целом, это то, что ты делаешь на строке выше. Но сейчас этот способ выглядит грязновато из-за того как реализован листнер на 50-ой строке;
2. Это передать в Адаптер ламбда выражение(этот способ **приоритетнее**).  Далее я буду описывать именно этот вариант.

sobolevkir replied:
А почему у меня слушатель организован был некорректно? Можешь чуть более подробно про это подсказать? Я вроде делал его как в теории, через функциональный интерфейс..

IliaAnK replied:
Пожалуй, я не совсем корректно выразился. Скорее я хотел сказать, что путь с функциональным интерфейсом более многословный и заставляет работать с nullable типом.

IliaAnK commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/TrackListAdapter.kt:
```diff
@@ -1,30 +1,47 @@
+        with(holder) {
+            bind(tracks[position])
+            binding.root.setOnClickListener {
+                if (clickDebounce()) {
```
Рекомендация. Шаг 2:
`clickDebounce()` стоит перенести в "SearchActivity", поскольку это лишняя ответственность для адаптера.
Всё что должен делать адаптер это передавать данные во вьюХолдеры и управлять их отображением

IliaAnK commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/TrackListAdapter.kt:
```diff
@@ -1,30 +1,47 @@
-            it.context.startActivity(playerIntent)
+        with(holder) {
+            bind(tracks[position])
+            binding.root.setOnClickListener {
```
Рекомендация. Шаг 3:
На основе первых двух шагов содержимое `setOnClickListener` будет выглядеть следующим образом:

```
binding.root.setOnClickListener {
                onItemClick(tracks[position])
            }
```

Далее расскажу откуда взялся `onItemClick`

IliaAnK commented on app/src/main/java/com/sobolevkir/playlistmaker/tracklist/TrackListAdapter.kt:
```diff
@@ -1,30 +1,47 @@
-                       private val onTrackClickListener: OnTrackClickListener? = null
+class TrackListAdapter(
+    private val tracks: MutableList<Track>,
+    private val onTrackClickListener: OnTrackClickListener? = null
```
Рекомендация. Шаг 4:
В первом шаге я говорил про лямбда выражение. В адаптер оно будет передаваться через конструктор и будет выглядеть так:

```
class TrackListAdapter(
    private val tracks: MutableList<Track>,
    private val onItemClick: (Track) -> Unit
)
```

sobolevkir replied:
Тогда интерфейс не нужен?

sobolevkir replied:
В теории просто был описан способ через функциональный интерфейс, почему это более грязный способ?

IliaAnK replied:
Да, если решишь заменить на лямбду, то интерфейс не нужен.

Ответ на второй вопрос оставил в треде "Шаг 1"

IliaAnK commented on app/src/main/java/com/sobolevkir/playlistmaker/SearchActivity.kt:
```diff
@@ -40,66 +40,58 @@ class SearchActivity : AppCompatActivity() {
 
         foundTracksAdapter = TrackListAdapter(tracksFound) {
             SearchHistory.addTrackToHistory(it)
             historyTracksAdapter.notifyDataSetChanged()
```
Рекомендация. Шаг 5:
Здесь нужно передать лямбда выражение в адаптер. Но сначала обсудим то что есть сейчас.

Метод `notifyDataSetChanged()` говорит о том что данные в адаптере изменились и ему нужно перерисовать все элементы. Его нужно использовать, когда ты обновил данные для списка. Он не имеет никакого отношения к передаче или получению данных из/в адаптер.

Теперь вернёмся к лямбде. Вместо `historyTracksAdapter.notifyDataSetChanged()` нужно написать следующее:

```
// Как писал до этого clickDebounce() стоит перенести в этот класс, поэтому ориентируемся на то, что он уже здесь есть
if (clickDebounce()) {
                val playerIntent = Intent(this, PlayerActivity::class.java)
                playerIntent.putExtra(CURRENT_TRACK, it)
                startActivity(playerIntent)
            }
```

sobolevkir replied:
Почему здесь не нужно обновлять данные в адаптере?
Данные же изменились как раз, т.к. происходит клик по треку в найденных вариантах, поэтому он добавляется в список истории..

IliaAnK replied:
Океей, я похоже проглядел, что это разные адаптеры. Подумал, что ты обновляешь `foundTracksAdapter `

sobolevkir replied:
Я убрал кстати это обновление адаптера истории, всё-равно оно каждый раз вызывается когда история показывается. Поэтому, как я понял, оно больше не требуется тут. Я правильно понимаю?

IliaAnK replied:
Да, правильно

IliaAnK commented on app/src/main/java/com/sobolevkir/playlistmaker/SearchActivity.kt:
```diff
@@ -40,66 +40,58 @@ class SearchActivity : AppCompatActivity() {
         }
-        trackSearchList.adapter = foundTracksAdapter
-
         historyTracksAdapter = TrackListAdapter(SearchHistory.historyTracks)
```
Рекомендация. Шаг 6:
Само собой здесь тоже нужно добавить лямбду.

```
historyTracksAdapter = TrackListAdapter(SearchHistory.historyTracks) {
    if (clickDebounce()) {
                val playerIntent = Intent(this, PlayerActivity::class.java)
                playerIntent.putExtra(CURRENT_TRACK, it)
                startActivity(playerIntent)
            }
}
```

IliaAnK replied:
Рекомендация. Шаг 7:
Как можно заметить, код внутри лямбды в основном повторяется и его можно вынести в отдельный метод.

sobolevkir replied:
тогда получается лямбда эта будет задублирована. А как нам говорили в теории, дублирования нужно стараться избегать, разве нет?

IliaAnK replied:
Да, всё правильно сказали.
Поэтому я и написал, что лямбду можно вынести в метод и использовать этот метод в двух местах.

sobolevkir committed:
Изменения:
1) Теперь при условии отстутствии записи в SharedPrefs тема приложения соответствует теме, установленной на самом устройстве.
Если же настройка в SharedPrefs уже есть, то тема соответствует ей.
2) Запуск PlayerActivity теперь передаётся в адаптер как лямбда-выражение.

sobolevkir committed:
Изменения:
1) Теперь при условии отстутствии записи в SharedPrefs тема приложения соответствует теме, установленной на самом устройстве.
Если же настройка в SharedPrefs уже есть, то тема соответствует ей.
2) Запуск PlayerActivity теперь передаётся в адаптер как лямбда-выражение.
3) Поправил экран плеера: теперь кнопка управления воспроизведением возвращается к состоянию Play по окончании воспроизведения.

sobolevkir committed:
Merge remote-tracking branch 'origin/dev' into dev

sobolevkir committed:
Добавил дженерик-тип <Track> в метод getParcelableExtra для получения данных о текущем треке в случае, когда API < 33.

IliaAnK commented:
Касаемо темы приложения, да ты всё правильно понял)
Рад, что у тебя получилось!

IliaAnK commented:
А по поводу `'getParcelableExtra(String!): T?' is deprecated. Deprecated in Java`

Здесь всё окей. Да, deprecated метод будет всегда подсвечиваться. Даже если ты обернул его в if с условием про версию андроида. От этого условия метод не перестаёт быть deprecated.

Молодец, что используешь свежий Api, там где это возможно!
